interfaces/cython/cantera/base.pyx

```diff
@@ -232,6 +232,21 @@ cdef class _SolutionBase:
         def __get__(self):
             return anymap_to_dict(self.base.parameters(True))
 
+    def update_user_data(self, data):
+        """
+        Add the contents of the provided `dict` as additional fields when generating
+        YAML phase definition files with `write_yaml` or in the data returned by
+        `input_data`. Existing keys with matching names are overwritten.
+        """
+        self.thermo.input().update(dict_to_anymap(data), True)
+
+    def clear_user_data(self):
+        """
+        Clear any user-provided auxillary input data, so that the data given by
+        `input_data` or `write_yaml` will only include values generated by Cantera.
+        """
+        self.thermo.input().clear()
```

Comment:
I don't see how this distinguishes between data that the user has added using `update_user_data` and the data coming from the YAML file that was used to create the `Solution`. Can you clarify that for me? Thanks :smile:

Reply:
It doesn't distinguish - both are erased here. But once the object's been constructed, Cantera doesn't need a pristine copy of the data from the input file, so clearing this doesn't have any noticeable effect besides removing the auxiliary data, since the necessary fields will be calculated through the calls to `getParameters`.

Reply:
So in principle, this does not support true round-tripping of an input file, right? If all the data are cleared, and any data computed by Cantera is not quite identical to the original file, the resulting generated file wouldn't be identical?

This may be a pedantic thing, but since the docstring says one thing, and the code actually does something else, I foresee potential problems as a result.

Reply:
Again, I think we're getting pretty far off the purpose of _this_ PR, as this basically all about features that were introduced in #984 if not earlier, but I can try to provide a bit of an explanation.

We don't keep track of how an object has changed since it was instantiated from the input file, so any time you ask for the `input_data` or call `write_yaml`, that's being calculated on the fly. I don't think there's anything that promises to precisely round-trip an input file, and there are currently a variety of ways in which that isn't currently possible. Some of these will probably be addressed relatively soon, like the serialization of the file-level metadata (as requested in #1013), while others may not (like retaining comments in YAML files). The only promise that I think we should be trying to make regarding serialization is that you can generate an input file that will recreate an object in its current state, and I think that is currently the case.

Reply:
> So in principle, this does not support true round-tripping of an input file, right? If all the data are cleared, and any data computed by Cantera is not quite identical to the original file, the resulting generated file wouldn't be identical?
>
> This may be a pedantic thing, but since the docstring says one thing, and the code actually does something else, I foresee potential problems as a result.

This is quite similar to what I had mentioned in one of my comments above (what does *input data* mean?). Either path forward is fine; I believe round-trip conversion would be simple to implement as stated earlier:

> The implementation would be quite simple, where all that is needed is a second `AnyMap` member variable that holds information added after the objects are instantiated, plus some tweaks for logic in some of the getters.

Reply:
I think it's fine if we consider changing/discussing the round-trip behavior out of scope for this PR. That said, I'd prefer if this docstring were a little more accurate to reflect what's actually happening. Since this very explicitly says that user-provided data will be cleared, it implies that's the only thing that's cleared. Maybe something like:

> Clear auxiliary input data, including user-provided values. When `write_yaml` or `input_data` are called, Cantera will regenerate the input data set to reproduce the current object state.

I'm not sure if this is accurate though; I'm thinking of the case where a user may have manually, e.g., changed a species thermo entry (for example, to debug a discontinuous polynomial) and _that_ data will be written out instead of the data that was originally used to create the `Solution()`. Is that correct?

Reply:
I had been using "auxiliary" to mean data that is not used by Cantera, so there is no auxiliary input data that isn't user-provided. I guess this would be more precise:
> Clear all saved input data, so that the data given by `input_data` or `write_yaml` will only include values generated by Cantera based on the current object state.

Reply:
That sounds great to me! Thanks!